### PR TITLE
[wip] Introducing Flytekit Rust entrypoint [flyrs]

### PR DESCRIPTION
**Pull request by kumare3**
## Tracking issue
 - Faster performance - faster startup. In limited testing more than 3.8x speedup
 - Smaller footprint - our goal is to remove object-store dependencies and replace with rust
 - Also remove dependency on grpc in python and only replace with rust

This also introduces rust more wholistically in flytekit

### CODE REVIEW

**Comment by kumare3** [comment]
@vlad-ivanov-name thank you for all the comments. As you can tell I am complete newbie and I will definitely try to adapt all your suggestions. If you think you can take over the PR and improve it - please feel free too :) I would love to hand it over

File: src/distribution.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+use log_derive::{logfn, logfn_inputs};
+use object_store;
+use object_store::{ObjectStore, parse_url};
+use object_store::path::Path;
+use tar::Archive;
+use url::Url;
+
+#[logfn_inputs(Info, fmt = "Downloading distribution from {} to {}")]
+#[logfn(ok = "INFO", err = "ERROR")]
+pub async fn download_unarchive_distribution(src: &Url, dst: &String) -> Result<(), Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (kumare3) [reply src/distribution.rs]: Done!

[the remark this answers is not in the document]
Comment (kumare3) [reply src/distribution.rs]: @vlad-ivanov-name can you check

File: flyrs_test/test.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export PYTHONPATH=`pwd`:$PYTHONPATH
+
+TESTDATA_PATH=`pwd`/testdata
+
+../target/debug/flyrs --inputs ${TESTDATA_PATH}/inputs.pb --output-prefix ${TESTDATA_PATH} --raw-output-data-prefix ${TESTDATA_PATH} --dynamic-addl-distro file://${TESTDATA_PATH}/schedule.tar.gz --dynamic-dest-dir . --resolver "flytekit.core.python_auto_container.default_task_resolver" -- task-module schedule task-name say_hello
+
+cmp testdata/outputs.pb testdata/expected_outputs.pb || echo -e "----------- Outputs file comparision failed! ----------"
```
[the remark this answers is not in the document]
Comment (kumare3) [reply flyrs_test/test.sh]: hmm i also wanted to test the cli, but let me learn

File: src/executor.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        let keys = modules.call_method0("keys").unwrap();
+        debug!("Python path: {:?}", path);
+        debug!("Python version: {:?}", version);
+        debug!("Python modules: {:?}", keys);
+    }
+}
+
+#[logfn_inputs(Info, fmt = "Invoking task with {}")]
+#[logfn(ok = "INFO", err = "ERROR")]
+pub async fn execute_task(args: &ExecutorArgs) -> Result<(), Box<dyn std::error::Error>>{
```
[the remark this answers is not in the document]
Comment (kumare3) [reply src/executor.rs]: thank you

File: rust/flyrs/src/executor.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+fn debug_python_setup(py: Python) {
+    if tracing::enabled!(tracing::Level::DEBUG) {
+        let sys = PyModule::import_bound(py, "sys").unwrap();
+        let path = sys.getattr("path").unwrap();
+        let version = sys.getattr("version").unwrap();
+        let modules = sys.getattr("modules").unwrap();
+        let keys = modules.call_method0("keys").unwrap();
+        debug!("Python path: {:?}", path);
+        debug!("Python version: {:?}", version);
+        debug!("Python modules: {:?}", keys);
```
[the remark this answers is not in the document]
Comment (kumare3) [reply rust/flyrs/src/executor.rs:65]: Wow you are indeed making me a better rust programmer - thank you appreciate it

**Comment by kdubovikov** [comment]
@kumare3 hi. Wanted to check when this is planned to be merged?

**Comment by kumare3** [comment]
This needs more work, especially tooling so - not yet planned but definitely later this year

**Comment by kdubovikov** [comment]
> This needs more work, especially tooling so - not yet planned but definitely later this year

@kumare3 is there anything that outside contributors can help with so that the release of this feature will be faster?

**Comment by kdubovikov** [comment]
@austin362667 , @kumare3, hi. Just a surface-level thought on the overall design. So, Flyte uses GRPC as a default transport, most likely due to being written in Go. Then, we inherited this transport in Python Entrypoint and flytekit. Afterward, community has realized that GRPC clients in Python is bound to some performance and stability issues in our specific use cases. So, the proposed solution is to create native GRPC clients in Rust and wrap them as Python packages with PyO3. Am I correct with this chain of reasoning?

@austin362667 By inspecting your PRs I see what you say when you mention that the task is not easy. You seem to employ quite deep workarounds for solving ref outliving.

If I am, then I wanted to ask if any alternative and probably simpler approaches were considered:

- What if Flyte API had a set of RESTful ports/endpoints in addition to the GRPC ones? Would it remove the performance issues in the Entrypoint? Do we know if the primary contributor to the slowness of the entrypoint is the GRPC client? I did not see much CPU-bound code in there. This approach would also remove GRPC dependencies from flytekit and make it more lightweight. `asyncio` could be used for communication to resolve any performance issues, and if the endpoint was over compressed HTTP/2 I suppose that the difference between GRPC and REST would be minimal. Just looking at the efforts needed to introduce Rust to the project along with additional GRPC client packaging using PyO3, I am wondering if simpler alternatives are viable? Are there any significant drawbacks that you see?
- Was it considered encapsulating all communication with Flyte APIs into a sidecar, making Task pods lighter. There would be no need for entrypoint this way, or at least it would be a minimal setup script.

I am not saying that the current solution is bad in any way, just asking questions from a fresh perspective.

**Comment by austin362667** [comment]
Actually I think you're right.

Refactoring gRPC into RESTful is another **huge** story, too. And, yes, it indeed mitigates the dependencies over Python `grpcio` and `protobuf`.

Let's say if we want to stay in gRPC architecture. It's true the transport layer is not CPU-bounded, and I've already enabling non-blocking network by leveraging `asyncio` during this [`run` / `register` asynchronously](https://github.com/flyteorg/flytekit/pull/2276). So performance is not an issue or bottleneck any more. Another reason to introduce Rust in the Flytekit remote client is to reduce the wheel size, thereby decreasing the Docker image pull wait time for Flytekit when starting a pod. Am I missing anything? @pingsutw 

IMO, a better use case for introducing Rust is rewriting the entry point, just like what @kumare3  is trying  here. Refactor this component might have much more potential for performance gains.

**Comment by kumare3** [comment]
Grpc is not used because of golang, but because of it being way faster than rest and type safety. Problem has been Grpc python support by Google has been not great. All the top projects in the world today use Grpc and I still prefer it. We are looking migrating to an http2 compliant implementation of Grpc - connect!

**Comment by austin362667** [comment]
Yeh, no doubt speed and type safety are the two most important reasons to use gRPC!

> 2. Was it considered encapsulating all communication with Flyte APIs into a sidecar, making Task pods lighter. There would be no need for entrypoint this way, or at least it would be a minimal setup script.

I'm not sure on this one. Can you elaborate more? And wdyt? @MortalHappiness

**Comment by kdubovikov** [comment]
> Grpc is not used because of golang, but because of it being way faster than rest and type safety. Problem has been Grpc python support by Google has been not great. All the top projects in the world today use Grpc and I still prefer it. We are looking migrating to an http2 compliant implementation of Grpc - connect!

@kumare3 understood. I can see reasoning behind this. I prefer strong typing in remote calls as well. However, if Python is the primary client language and GRPC support there is not great, is it a solid reason for introducing another language (and a quite tricky one) to the project?

In case of the Endpoint, why not implement it in Go, for example? Is it a drastic difference between calling the Task function from Go, compared to using PyO3 in Rust? If Go was used, probably we would achieve the same goals along with reducing the maintainability pressure by introducing a new language to the project.

And feel free to tell me if I am not making much sense, I will stop debating on this particular point :)

> Another reason to introduce Rust in the Flytekit remote client is to reduce the wheel size, thereby decreasing the Docker image pull wait time for Flytekit when starting a pod.

@austin362667 this is a good point indeed. Do we know by how much the wheel size is reduced?

**Comment by MortalHappiness** [comment]
> 2. Was it considered encapsulating all communication with Flyte APIs into a sidecar, making Task pods lighter. There would be no need for entrypoint this way, or at least it would be a minimal setup script.

@kdubovikov @austin362667 

Although I do not have much context with the Rust gRPC implementation, from a K8s perspective, I do not think this is correct. First off, sidecar container also run in the same Pod with the main container, so the task pods will not be lighter. Besides, I don't think it is easier to implement it as a sidecar container. As far as I know, the main communication ways between sidecar and main container is via network or shared volumes, so the sidecar container must be either consistently watch for file system changes or serves as a translator between the main container and the Flyte API. In either way the implementation will not be easier.

Feel free to tell me if you have a better implementation for the sidecar pattern. Maybe I misunderstand something.

**Comment by kdubovikov** [comment]
> > 2. Was it considered encapsulating all communication with Flyte APIs into a sidecar, making Task pods lighter. There would be no need for entrypoint this way, or at least it would be a minimal setup script.
> 
> @kdubovikov @austin362667
> 
> Although I do not have much context with the Rust gRPC implementation, from a K8s perspective, I do not think this is correct. First off, sidecar container also run in the same Pod with the main container, so the task pods will not be lighter. Besides, I don't think it is easier to implement it as a sidecar container. As far as I know, the main communication ways between sidecar and main container is via network or shared volumes, so the sidecar container must be either consistently watch for file system changes or serves as a translator between the main container and the Flyte API, by capturing the network requests from the main container and translate it to gRPC. In either way the implementation will not be easier.
> 
> Feel free to tell me if you have a better implementation for the sidecar pattern. Maybe I misunderstand something.

@MortalHappiness I meant that the Task container will get lighter and possibly won't need extra GRPC dependencies, and the sidecar can act as a bridge and use any more efficient implementation that we consider to be faster. This way the task container won't need much except from the user code itself, and sidecar would take care of handling all of the platform functionality. However, I agree that adding a sidecar is not contributing much to the entrypoint performance by itself, and the problem of communication between sidecar and task containers still must be solved. The key to this issue is optimizing the entrypoint cold start time, and sidecar is more about decoupling one from another. So yes, let's dismiss the sidecar thread for now.

It's just that using PyO3 and Rust as a workaround for standard Google GRPC clients in Python is not looking as a very common or widely accepted pattern that I have seen in other codebases (especially those that are not primarily in Rust), so I ask if everyone involved are sure that introducing Rust to the project is the best possible solution considering all tradeoffs. Maybe I am not aware of other roadmap points that will broaden the usage of Rust for different use cases as well. Again, not saying that using Rust for creating Python GRPC client packages or entrypoint script is wrong in any way, just mentioning that it adds a whole new layer of maintainability in the long term.

The second usecase is to use Rust for the Entrypoint script, but I am not entirely sure if it will be much different than using Go for the same: run task setup code and then run and monitor the Python process, communicating back to Flyte when neccessary.

**In short**
* Does Flyte has far-reaching plans to use Rust in the future for any other cases than Entrypoint?
* If no, why won't we write entrypoint in Go, which is already the primary language of the platform and should give us performance benefits as well.

**Comment by kumare3** [comment]
@kdubovikov grpc is not used at runtime. This communication happens through blob store. 
The decision on a new language is dependent on what we can achieve with the new language. 
The reason for rust is the amazing benefit of interoperability with many different languages. Goal is to build a common rust runtime that allows any language sdk.

As for the sidecar pattern - it is already supported in Flyte through raw container tasks. There are many disadvantages with it too. Check it out

**Comment by kdubovikov** [comment]
> @kdubovikov grpc is not used at runtime. This communication happens through blob store. The decision on a new language is dependent on what we can achieve with the new language. The reason for rust is the amazing benefit of interoperability with many different languages. Goal is to build a common rust runtime that allows any language sdk.
> 
> As for the sidecar pattern - it is already supported in Flyte through raw container tasks. There are many disadvantages with it too. Check it out

Thanks for the context, @kumare3 . `Go` would not solve the interoperability for sure, if that's our goal. Then Rust it is 🦀 , closing this discussion.

My first question remains: is there anything that can be done by outside contributors to speed up the release of this feature?

**Comment by kumare3** [comment]
@kdubovikov and others, we will be sharing an updated on Flyte2.0, which is a complete reimagining of the system, and has a lot of RUST power
